src/packaging/markers.py

```diff
@@ -50,6 +51,89 @@ class UndefinedEnvironmentName(ValueError):
     """
 
 
+class Environment(TypedDict, total=False):
+    implementation_name: str
+    """The implementation's identifier, e.g. ``'cpython'``."""
+
+    implementation_version: str
+    """
+    The implementation's version, e.g. ``'3.13.0a2'`` for CPython 3.13.0a2, or
+    ``'7.3.13'`` for PyPy3.10 v7.3.13.
+    """
+
+    os_name: str
+    """
+    The value of :py:data:`os.name`. The name of the operating system dependent module
+    imported, e.g. ``'posix'``.
+    """
+
+    platform_machine: str
+    """
+    Returns the machine type, e.g. ``'i386'``.
+
+    An empty string if the value cannot be determined.
+    """
+
+    platform_release: str
+    """
+    The system's release, e.g. ``'2.2.0'`` or ``'NT'``.
+
+    An empty string if the value cannot be determined.
+    """
+
+    platform_system: str
+    """
+    The system/OS name, e.g. ``'Linux'``, ``'Windows'`` or ``'Java'``.
+
+    An empty string if the value cannot be determined.
+    """
+
+    platform_version: str
+    """
+    The system's release version, e.g. ``'#3 on degas'``.
+
+    An empty string if the value cannot be determined.
+    """
+
+    python_full_version: str
+    """
+    The Python version as string ``'major.minor.patchlevel'``.
+
+    Note that unlike the Python :py:data:`sys.version`, this value will always include
+    the patchlevel (it defaults to 0).
+    """
+
+    platform_python_implementation: str
+    """
+    A string identifying the Python implementation.
+
+    Currently, the following implementations are identified: ``'CPython'`` (C
+    implementation of Python), ``'IronPython'`` (.NET implementation of Python),
+    ``'Jython'`` (Java implementation of Python), ``'PyPy'`` (Python implementation of
+    Python).
+    """
+
+    python_version: str
+    """The Python version as string ``'major.minor'``."""
+
+    sys_platform: str
+    """
+    This string contains a platform identifier that can be used to append
+    platform-specific components to :py:data:`sys.path`, for instance.
+
+    For Unix systems, except on Linux and AIX, this is the lowercased OS name as
+    returned by ``uname -s`` with the first part of the version as returned by
+    ``uname -r`` appended, e.g. ``'sunos5'`` or ``'freebsd8'``, at the time when Python
+    was built.
+    """
+
+    extra: Optional[str]
```

Comment:
I was thinking about this over the weekend and I realized this is represents something being wrong. Either `evaluate()` has a bad type hint because it says it takes `dict[str, str]` and yet it checks if the `"extra"` key is `None`, or we don't want this key since it doesn't type against that pre-existing parameter type and it isn't accurate for what `default_environment()` returns.

IOW we need to decide it this `TypedDict` represents what `default_environment()` returns, or the dict of potential marker values that `evaluate()` works with? Any opinions?

If it's the latter and it type checks w/ arbitrary keys, I think we can revert the change I requested. 😅

Reply:
> `TypedDict` represents what `default_environment()`

I think this is more accurate. With that assumption, I added a `cast` in `Marker.evaluate`: eaa4322.

Reply:
If it represents the return type, does that mean the `total` part should be removed?

Reply:
Yeah, that makes sense: 08fc194fcf0fcd08b5296839ab7b0af83be81c98